extension/icu/icu-timebucket.cpp

```diff
@@ -0,0 +1,606 @@
+				    });
+			} else {
+				TernaryExecutor::ExecuteWithNulls<interval_t, timestamp_t, timestamp_t, timestamp_t>(
+				    bucket_width_arg, ts_arg, origin_arg, result, args.size(),
+				    [&](interval_t bucket_width, timestamp_t ts, timestamp_t origin, ValidityMask &mask, idx_t idx) {
+					    return OriginTernaryOperator::Operation<interval_t, timestamp_t, timestamp_t, timestamp_t>(
+					        bucket_width, ts, origin, mask, idx, calendar);
+				    });
+			}
+		}
+	} else {
+		TernaryExecutor::ExecuteWithNulls<interval_t, timestamp_t, timestamp_t, timestamp_t>(
+		    bucket_width_arg, ts_arg, origin_arg, result, args.size(),
+		    [&](interval_t bucket_width, timestamp_t ts, timestamp_t origin, ValidityMask &mask, idx_t idx) {
+			    return OriginTernaryOperator::Operation<interval_t, timestamp_t, timestamp_t, timestamp_t>(
+			        bucket_width, ts, origin, mask, idx, calendar);
+		    });
+	}
+}
+
+void ICUTimeBucket::ICUTimeBucketTimeZoneFunction(DataChunk &args, ExpressionState &state, Vector &result) {
+	D_ASSERT(args.ColumnCount() == 3);
+
+	auto &func_expr = (BoundFunctionExpression &)state.expr;
+	auto &info = (BindData &)*func_expr.bind_info;
+	CalendarPtr calendar_ptr(info.calendar->clone());
+	auto calendar = calendar_ptr.get();
+
+	auto &bucket_width_arg = args.data[0];
+	auto &ts_arg = args.data[1];
+	auto &tz_arg = args.data[2];
+
+	if (bucket_width_arg.GetVectorType() == VectorType::CONSTANT_VECTOR &&
+	    tz_arg.GetVectorType() == VectorType::CONSTANT_VECTOR) {
+		if (ConstantVector::IsNull(bucket_width_arg) || ConstantVector::IsNull(tz_arg)) {
+			result.SetVectorType(VectorType::CONSTANT_VECTOR);
+			ConstantVector::SetNull(result, true);
+		} else {
+			interval_t bucket_width = *ConstantVector::GetData<interval_t>(bucket_width_arg);
+			SetTimeZone(calendar, *ConstantVector::GetData<string_t>(tz_arg));
+			if (bucket_width.months == 0 && Interval::GetMicro(bucket_width) > 0) {
+				BinaryExecutor::Execute<interval_t, timestamp_t, timestamp_t>(
+				    bucket_width_arg, ts_arg, result, args.size(), [&](interval_t bucket_width, timestamp_t ts) {
+					    return TimeZoneWidthLessThanDaysBinaryOperator::Operation<interval_t, timestamp_t, timestamp_t>(
+					        bucket_width, ts, calendar);
+				    });
+			} else if (bucket_width.months > 0 && bucket_width.days == 0 && bucket_width.micros == 0) {
+				BinaryExecutor::Execute<interval_t, timestamp_t, timestamp_t>(
+				    bucket_width_arg, ts_arg, result, args.size(), [&](interval_t bucket_width, timestamp_t ts) {
+					    return TimeZoneWidthMoreThanMonthsBinaryOperator::Operation<interval_t, timestamp_t,
+					                                                                timestamp_t>(bucket_width, ts,
+					                                                                             calendar);
+				    });
+			} else {
+				TernaryExecutor::Execute<interval_t, timestamp_t, string_t, timestamp_t>(
```

Comment:
Error condition?

Reply:
To avoid throwing error when some argument is null, I call `TernaryExecutor::Execute` that has null check.